rust/lance-index/src/vector/graph/memory.rs

```diff
@@ -58,22 +58,25 @@ impl VectorStorage for InMemoryVectorStorage {
     fn dist_calculator(&self, query: &[f32]) -> Box<dyn DistCalculator> {
         Box::new(InMemoryDistanceCal {
             vectors: self.vectors.clone(),
-            query: query.to_vec(),
+            query: query.as_ptr(),
             metric_type: self.metric_type,
         })
     }
 }
 
 struct InMemoryDistanceCal {
     vectors: Arc<MatrixView<Float32Type>>,
-    query: Vec<f32>,
+    query: *const f32,
```

Comment:
Can we use `&[f32]` with lifetime? Using raw pointer means that the lifetime of query is not guaranteed when the `distance()` method is called.

Reply:
i tried this but found that the `dist_calc` is created by the trait method of `VectorStorage`, so it's impossible to add lifetime for this.

any better way?

Reply:
if we don't care about locality
`Arc<[f32]>` works

Reply:
We should probably make queries some sort of `Arc<[T]>` early in the query, so it's cheaply clone-able.

Reply:
a bit more pedantic. What we want is exactly this: https://github.com/pytorch/pytorch/blob/main/c10/util/intrusive_ptr.h

but I'm not sure if there is an equivalent and easily implementable in rust

Reply:
the vector in `Query` is `Arc<dyn Array>`, changed the type from `&[f32]` to `ArrayRef`